add wlr-foreign-toplevel-management protocol to enumerate and control desktop surfaces

ammen99 opened the pull request:
I'm opening this PR to discuss what we need to have in this protocol. I'm not very familiar with writing wayland protocols "the proper way", so any recommendations are very welcome :)

Things I'm not sure about:

1. Thumbnails - some prior art like wayland-wall required the compositor draw the view in a specified place on the screen. However, I think this is better solved if the app could create screenshots of individual windows.

2. Minimizing. When the compositor supports more advanced window minimize animations, it usually needs a place on the screen to minimize towards (think of the magic lamp effect in compiz).

I think this fixes #14, but I'm not sure whether consensus was reached if we should include window positions or not.

agx commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,144 @@
+
+    <event name="title">
+      <description summary="iitle change">
+        This event is emitted whenever the title of the window changes
```
title?

agx commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,144 @@
+
+  <interface name="zwlr_window_manager_v1" version="1">
+    <description summary="list and control opened apps">
+      The purpose of this protocol is to enable the creation of taskbars
```
Is this also about controlling or just about listing?

ammen99 replied:
also controlling, there are focus/maximize/minimize/close requests

agx commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,144 @@
+        This event is emitted whenever a new toplevel window is created.
+      </description>
+      <arg name="window" type="new_id" interface="zwlr_window_v1"/>
+    </event>
```
Is the list of current windows supposed to be sent when the client binds to zwlr_window_manager_v1? If so the protocol should specify it.

ammen99 replied:
yes, it is so, will do it.

agx commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,144 @@
+      <arg name="app_id" type="string"/>
+    </event>
+
+    <event name="enter_output">
```
Should that be "entered_output" to match the below "maximized"?

ammen99 replied:
no, it is more like the `wl_surface.enter` event

SirCmpwn replied:
Turning that question around: should `maximized` be called `maximize`?

SirCmpwn replied:
Also, I'd prefer `output_leave` and `output_enter` over the reverse

ammen99 commented:
@agx we might want to indicate that the window is xwayland and give access to the window ID so that the client can get the app icon. Aside from that I don't see any reason to expose the "type" of the window. 
Something I wondered, when I wrote this I was focused on "regular" windows. Is it clear panels/backgrounds are excluded?

emersion commented:
Consider mirroring xdg-toplevel and using xdg-toplevel's vocabulary.

> Is it clear panels/backgrounds are excluded?

IMHO yes.

>Thumbnails - some prior art like wayland-wall required the compositor draw the view in a specified place on the screen. However, I think this is better solved if the app could create screenshots of individual windows.

I don't agree. I think adding a request creating `wl_buffer`s that mirror other views and can be attached to surfaces is a better approach.

>Minimizing. When the compositor supports more advanced window minimize animations, it usually needs a place on the screen to minimize towards (think of the magic lamp effect in compiz).

If included, this should be optional.

SirCmpwn commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,148 @@
+    <request name="destroy" type="destructor"/>
+  </interface>
+
+  <interface name="zwlr_window_v1" version="1">
```
I think this language is a bit too generic. Imagine this eventually becomes stable and we're looking at a time named `xdg_window`. I think maybe `window_handle` or `foreign_toplevel` would be better.

emersion commented:
>Okay, so that includes splitting the maximize/minimize request into set_maximized and unset_maximized, and sending the states as an array? (if I got you correctly)

I'd also rename "window" → "toplevel". Windows can be popups (and your clipboard too! and your screen too!) in X11, sticking to the xdg-shell vocabulary makes things clearer.

SirCmpwn commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,148 @@
+        is on and whenever it is moved to a different virtual desktop later.
+      </description>
+      <arg name="desktop" type="string"/>
+    </event>
```
This probably belongs in a separate protocol. It's a subtle issue.

https://github.com/swaywm/wlr-protocols/issues/10

ammen99 replied:
okay, will remove it for now

SirCmpwn commented on unstable/wlr-window-list.xml:
```diff
@@ -0,0 +1,148 @@
+        Request that this window is focused. However, there is no guarantee
+        the window will be actually focused.
+      </description>
+    </request>
```
Should we call this "activate"? Also, there's no event which tells you if the window is focused or not.

ammen99 replied:
hmm, yes, this will better fit with xdg-shell wording

ammen99 commented:
I updated the protocol to use `foreign_toplevel` and also addressed the other comments. I also added the set_rectangle request, which is used to specify where on the screen the toplevel is represented

agx commented:
Since activation is part of the protocol should raising the window to the top of the stack also be or is this considered to be compositor specific (like raise when activated if necessary)?

ammen99 commented:
> For e.g. terminating applications it would be good to have the pid of the client that owns the surface. Should there be a request for that too?

IMO force-kill belongs to the compositor, not the dock/taskbar/whatever

agx commented:
> IMO force-kill belongs to the compositor, not the dock/taskbar/whatever

Imagine a touch device where the "taskbar" is fullscreen and lists all running applications to switch to them. Kill/Forcekilling fits in there nicely. Same holds for a HUD on the desktop. An alternative would be to have "kill client matching toplevel" request so we don't need to expose the pid.

ammen99 commented:
@agx I was thinking of desktop usage, where you usually get "force-kill or wait" dialogue. But indeed on a smartphone it would make more sense to just kill it. But isn't the task switcher built-in into the compositor in this case?

Ongy commented:
PID doesn't necessarily make sense.
And when the socket was created with `socketpair()` and passed to the client, I don't think it's necessarily possible to get it.

So the way to go should be over the compositor that simply disconnects the client as worst case measure (or tries to crawl /proc/ to get the pid, but that's not very nice either)

SirCmpwn commented:
I think the compositor can use xdg-shell ping shortly after it's been asked to close a surface, and if the client doesn't respond it can show its own UI for force killing the application.

agx commented:
Well that implies that the compositor wants to do UI as well (with all implications like pulling in a UI toolkit etc). I don't think the protocol should mandate that. Since the compositor can't be restarted I'd rather have these things outside of it. That doesn't mean that it has to be in *this* protocol but since maximize and minimize is in there too it fits.

SirCmpwn commented:
I think it's giving the client too much power. A compositor that doesn't want to make the UI for that could just force kill clients when a second close request is made after enough time without responding to a ping. It could also change the title to "[Unresponsive] $title" or something, too.

ammen99 commented:
@agx if you want force-kill, then why not bake this into the compositor? I think desktop compositors will implement such a popup in some way anyway, and for touch-based devices, just kill after a short timeout by default.

agx commented:
Having the compositor run the same UI toolkit than the shell makes little sense (and since it's not restartable I want as little baggage in the compositor as possible). Doing that in process is just not necessary (neither on the desktop nor on mobile devices).

I also don't buy the "power" argument since clients allowed to certain protocols have lots of power already (like faking arbitrary appliacations by using screen shot and layer surface). As I said it doesn' t have to go into this protocol, I can always add a separate one on top.

SirCmpwn commented:
Alright, separate protocol seems fine.

agx commented:
Rereading this: the last comments only focus on force killing but I'm also talking about terminating application (SIGTERM) e.g. the shell displays the toplevels and you swipe up to close the application or like gnome-shell displays a little "x" in the top-right corner of the overview that can be clicked to close the application. 
(Just for completeness so it doesn't get lost when I pick it up again)

emersion commented:
>terminating application (SIGTERM) e.g. the shell displays the toplevels and you swipe up to close the application or like gnome-shell displays a little "x" in the top-right corner of the overview that can be clicked to close the application.

This is done via a "close" event in xdg-shell (and a similar mechanism on X11 when the client supports it). This is not done by sending a signal to the process.

agx commented:
@emersion good point. My point was basically that it's not only about "force-kill" but a also about application windows. Using the means of the underlying protocol (e.g. xdg-shell) is certainly better than plain unix signals.

So maybe plain "close" fits in here while force-kill does not?

ammen99 commented:
But @agx is right, such a request (like xdg-shell close) belongs to this protocol, and I thought I had included it. Maybe I deleted it accidentally, will fix that.

ammen99 commented:
Is the newly added close-request satisfactory?

agx commented:
@ammen99 looks good to me!

SirCmpwn commented:
LGTM

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,190 @@
+      <arg name="toplevel" type="new_id" interface="zwlr_foreign_toplevel_v1"/>
+    </event>
+
+    <request name="destroy" type="destructor"/>
```
This should have a description explaining what happens to objects created via `toplevel_created`.

It's tricky to design an interface that doesn't leak.

ammen99 replied:
do we need to explain that after I changed the proto to use client-generated objects?

emersion replied:
Yes.

ammen99 commented:
I do hope the protocol is in a better shape now, thank you all for the patience.

ammen99 commented:
Yet another change, inspired by a discussion in #wayland:

> use server-side created resources, add a request on the factory to ask the compositor to stop sending resources in events

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<protocol name="wlr_foreign_toplevel_list_v1">
```
The name should be `wlr_foreign_toplevel_list_unstable_v1`

ammen99 replied:
I guess I should rename the file, too?

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+      toplevel window will be sent via the toplevel_created signal.
+    </description>
+
+    <event name="toplevel_created">
```
This can just be named `toplevel` to follow the Wayland naming conventions (e.g. `wl_output` has a `scale` event, `wl_data_device` has `data_offer`...)

ammen99 replied:
okay

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+      <description summary="stop sending events">
+        Indicates the client no longer wishes to receive events for new toplevels.
+        However the compositor may emit further toplevel_created events, until
+        the finished event is emitted.
```
Add that the client cannot send more requests after this one

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+        This event indicates that the compositor is done sending events to the
+        zwlr_foreign_toplevel_manager_v1. The server will destroy the resource
+        immediately after sending this request. The client should still call
+        zwlr_foreign_toplevel_v1_destroy() as a response without sending any
```
This is not really a protocol concern, and mentioning a function call ties this protocol to libwayland. I'd just say something among the lines of "the object is not valid anymore and any resource associated to the object should be released".

ammen99 replied:
I also thought about that, will do.

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+      <entry name="activated" value="2" summary="the toplevel is active"/>
+    </enum>
+
+    <event name="state_changed">
```
This event can just be named `state`.

I wonder if we want a way to atomically commit state (like the `commit` event for `wl_surface`, or the `configure` event for `xdg_toplevel`). This might be useful if someone wants to design new protocols on top of this one to retrieve more properties.

ammen99 replied:
Well, all states are sent together anyway, this might make sense for the requests (atomicity). But they might be ignored, does it still make sense to be atomic?

ammen99 replied:
this is still unresolved, not sure how to make github display it properly

emersion commented on unstable/wlr-foreign-toplevel-list.xml:
```diff
@@ -0,0 +1,217 @@
+        This event means the toplevel has been destroyed. It is guaranteed there
+        won't be any more events for this zwlr_foreign_toplevel_v1. The client
+        should destroy the zwlr_foreign_toplevel_v1 object without sending any
+        further requests.
```
>without sending any further requests

This is racy. Is this a protocol error?

You might want to just say something like "the object becomes inert and any further request will be ignored" instead.

emersion replied:
After the server sends this event, it might receive more client requests (that the client has sent before receiving the event). The server can't guess whether the client has sent those before receiving the event or after (so making this a protocol error is racy).

ammen99 replied:
I see what you mean.

ammen99 commented:
later as in a few hours

SirCmpwn commented:
:+1:

ammen99 commented:
I have added a simple done event, because that's how it is called in xdg-output and seems closest to our purpose.

I have several questions:

1. Do we list all events that are included in the done event (output_enter/leave, state ..)?
2. destroyed event shouldn't be included in any case, right? I don't think it makes sense for any kind "atomicity" when the destroyed event is concerned
3. Do we require that done was sent before the destroyed?

ammen99 commented:
bump? anything else I should consider?

SirCmpwn commented:
You should probably start writing an example client and server implementation to prove this out.

ammen99 commented:
Sorry for the long delay, but I'had to focus on other things. I updated the protocol and the implementation to include the proposed seat argument, thanks @nyorain

ammen99 commented:
bump?

emersion commented on unstable/wlr-foreign-toplevel-unstable-v1.xml:
```diff
@@ -0,0 +1,229 @@
+      letting them request certain actions on them, like maximizing, etc.
+
+      After a client binds the zwlr_foreign_toplevel_manager_v1, each opened
+      toplevel window will be sent via the toplevel_created signal.
```
toplevel_created → toplevel

signal → event

emersion commented on unstable/wlr-foreign-toplevel-unstable-v1.xml:
```diff
@@ -0,0 +1,229 @@
+      </description>
+    </request>
+
+    <event name="finished">
```
This event should be marked as destructor.

emersion commented on unstable/wlr-foreign-toplevel-unstable-v1.xml:
```diff
@@ -0,0 +1,229 @@
+      <arg name="height" type="int"/>
+    </request>
+
+    <event name="destroyed">
```
destroyed → closed?

emersion replied:
The fact that you need to call `destroy` on a `destroyed` object is confusing.

ammen99 replied:
maybe "done" then?

emersion replied:
Nah, `done` is used as the compositor-side version of `commit`. See for instance `wl_output`.

emersion commented:
Overall, I'm not quite sure I like this design. I expect this protocol to be used for things like screenshooting and screencasting too, and I don't like how things like `set_minimized` are exposed here.

I'd prefer having two protocols. One protocol which just lists toplevels, and another one which extends foreign toplevels with requests like `set_minimized`.

SirCmpwn commented:
This protocol is designed to serve taskbars more so than screenshots. Even so, I like your proposed split.

ammen99 commented:
As discussed in IRC, this protocol needs to be renamed. Does the name `wlr-toplevel-management` (or `wlr-foreign-toplevel-management`, but I find that a bit too long) make sense? Then the `wlr_toplevel` interface to `wlr-toplevel-handle`?

If later a generic toplevel listing protocol is implemented, it could be called `wlr-toplevel-list` with `wlr-toplevel` representing the individual toplevel, without naming clashes. Plus, a 'handle' is AFAIK generally something you can use to control something, making it a good choice for this protocol.

ddevault commented:
I like `wlr-foreign-toplevel-management`.

emersion commented on unstable/wlr-foreign-toplevel-management-unstable-v1.xml:
```diff
@@ -0,0 +1,230 @@
+        However the compositor may emit further toplevel_created events, until
+        the finished event is emitted.
+
+        The client must not send any more requests after this one.
```
We should add an error code for this.

ammen99 replied:
it should be noted that we (in wlroots) destroy the resource server-side immediately when receiving this. How is the error handled in this case, won't libwayland drop the request or error the client?

emersion replied:
Hmm, you're right. It's probably fine then.

emersion commented on unstable/wlr-foreign-toplevel-management-unstable-v1.xml:
```diff
@@ -0,0 +1,230 @@
+        considered.
+
+        The dimensions are given in surface-local coordinates.
+        Setting width=height=0 removes the already-set rectangle.
```
Add an error for invalid values.

emersion commented:
Thanks!
